ingestion_server/ingestion_server/state.py

```diff
@@ -55,31 +57,44 @@ def register_indexing_job(worker_ips, target_index):
         db["worker_statuses"] = worker_statuses
         db["start_time"] = datetime.datetime.now()
         db["target_index"] = target_index
+        db["task_id"] = task_id
         return True
 
 
-def worker_finished(worker_ip):
+def worker_finished(worker_ip, error):
     """
     The scheduler received a notification indicating an indexing worker has
     finished its task.
     :param worker_ip: The private IP of the worker.
-    :return: The target index if all workers are finished, else False.
+    :param error: Whether this worker had an error during processing.
+    :return: A dict containing the target index, task_id, and the percent of
+    workers that have completed.
     """
     with FileLock(lock_path), shelve.open(shelf_path, writeback=True) as db:
         try:
             _ = db["worker_statuses"][worker_ip]
-            db["worker_statuses"][worker_ip] = WorkerStatus.FINISHED
+            db["worker_statuses"][worker_ip] = (
+                WorkerStatus.FINISHED if not error else WorkerStatus.ERROR
+            )
             log.info(f"Received worker_finished signal from {worker_ip}")
         except KeyError:
             log.error(
                 "An indexer worker notified us it finished its task, but "
                 "we are not tracking it."
             )
+        total_workers = len(db["worker_statuses"])
+        completed_workers = 0
         for worker_key in db["worker_statuses"]:
-            if db["worker_statuses"][worker_key] == WorkerStatus.RUNNING:
+            status = db["worker_statuses"][worker_key]
+            if status == WorkerStatus.RUNNING:
                 log.info(f"{worker_key} is still indexing")
-                return False
-        return db["target_index"]
+            elif status == WorkerStatus.FINISHED:
+                completed_workers += 1
```

Comment:
If I'm understanding this correctly: A worker that failed will report it encountered an error. It's status will be `WorkerStatus.ERROR`, so when these results are accumulated the completed count will _always_ be one less. Then, even if `active_workers` is false and thus `active` is false, `error` will still be true because `percent_completed < 100`?

Reply:
Yep. That being said I now think there's a problem -- which is that `active_workers` is not set to false until the `go_live` step completes, which will not happen if a worker has errored.

I think at minimum we should turn `active_workers` off when all indexers have completed, [_if_ any of them have errored.](https://github.com/WordPress/openverse-api/pull/593#discussion_r838989489)

This definitely feels a bit shaky around the error states and I would very much prefer to get this into Airflow for better tracking 😓 I think this should at least get us to a better place than we are currently.

Reply:
> This definitely feels a bit shaky around the error states and I would very much prefer to get this into Airflow for better tracking sweat I think this should at least get us to a better place than we are currently.

I'm so with you :sweat_smile: This just needs to be "good enough" to work on its own _most of the time_ until we can provide better support for it in Airflow